File: amrutha-academy-backend/lib/utils/apiResponse.ts
import { ApiResponse, PaginationResponse } from '@/types';

export function createSuccessResponse<T>(
  data: T,
  statusCode: number = 200,
  message?: string[]
): ApiResponse<T> {
  return {
    statusCode,
    message,
    data,
  };
}

export function createErrorResponse(
  error: string,
  statusCode: number = 400
): ApiResponse<null> {
  return {
    statusCode,
    error,
    data: null,
  };
}

export function createPaginatedResponse<T>(
  data: T[],
  page: number = 1,
  limit: number = 10,
  total?: number
): ApiResponse<T[]> {
  const totalItems = total ?? data.length;
  const totalPages = Math.ceil(totalItems / limit);
  
  const pagination: PaginationResponse = {
    page,
    limit,
    total: totalItems,
    totalPages,
  };

  return {
    statusCode: 200,
    data,
    pagination,
  };
}

File: amrutha-academy/lib/data/models/api_response.dart
class ApiResponse<T> {
  final int statusCode;
  final List<String>? message;
  final String? error;
  final T? data;
  final PaginationData? pagination;

  ApiResponse({
    required this.statusCode,
    this.message,
    this.error,
    this.data,
    this.pagination,
  });

  factory ApiResponse.fromJson(
    dynamic json,
    T Function(dynamic) fromJsonT,
  ) {
    final Map<String, dynamic> jsonMap = json as Map<String, dynamic>;
    return ApiResponse<T>(
      statusCode: jsonMap['statusCode'] ?? 200,
      message: jsonMap['message'] != null
          ? List<String>.from(jsonMap['message'])
          : null,
      error: jsonMap['error'],
      data: jsonMap['data'] != null ? fromJsonT(jsonMap['data']) : null,
      pagination: jsonMap['pagination'] != null
          ? PaginationData.fromJson(jsonMap['pagination'] as Map<String, dynamic>)
          : null,
    );
  }

  bool get isSuccess => statusCode >= 200 && statusCode < 300;
}

class PaginationData {
  final int page;
  final int limit;
  final int total;
  final int totalPages;

  PaginationData({
    required this.page,
    required this.limit,
    required this.total,
    required this.totalPages,
  });

  factory PaginationData.fromJson(Map<String, dynamic> json) {
    return PaginationData(
      page: json['page'] ?? 1,
      limit: json['limit'] ?? 10,
      total: json['total'] ?? 0,
      totalPages: json['totalPages'] ?? 0,
    );
  }
}

